### PR TITLE
US318359: Juned & Charlie: Added validation for cvv only functionality with unit tests

### PR DESCRIPTION
**Pull request by juned-miah**

### CODE REVIEW

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        verify(cardListener).onUpdate(cvvView, true)
+        verify(cardListener).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice and onUpdateLengthFilter event thrice - when cvv is partially valid and onUpdateCVV is called`() {
+        givenCvvIsOnFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV().onUpdateCVV(cvvValue)
```
Comment (ochalet-wp) [inline checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: Would the following help to test only the events triggered by the call to onUpdateCVV? (which is what we test here)

```
givenCvvIsOnFocus()
givenCvvIsPartiallyValid()
val  checkout =  createNewAccessCheckoutCVV()
reset(cardListener)

checkout.onUpdateCVV(cvvValue)
```

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        accessCheckoutCVV.cardValidator = null
+
+        accessCheckoutCVV.onEndUpdateCVV(cvvValue)
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any update events if there is no cardListener set and cvv is on focus`() {
```
Comment (ochalet-wp) [inline checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: Do you think we need to keep this test? This is essentially testing that if there's not listener then the listener is not called, right?

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = null
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any onUpdateLengthFilter events when cvvLengthFilter is null`() {
```
Comment (ochalet-wp) [inline checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: [Can merge] - I've seen that in the code the `cvvLengthFilter` is always set when the `cardValidator` is set (due to the behaviour of the `AccessCheckoutCardDefaultFactory`). So I presume that this covers the case where the client provides their own implementation of the `CardFactory`, right?

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+        given(factory.getCVVLengthFilter(cardValidator, null)).willReturn(null)
+
+        createNewAccessCheckoutCVV()
+
+        verifyZeroInteractions(cvvLengthFilter)
+        verify(cardListener, times(0)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    private fun createNewAccessCheckoutCVV(): AccessCheckoutCVV {
```
Comment (ochalet-wp) [inline checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: [Can merge] - Can be moved after the last test below

File: checkout/access-checkout-android/src/main/java/com/worldpay/access/checkout/AccessCheckoutCVV.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+        return cardValidator?.let {
+            val cvvResult = it.validateCVV(cvv, null)
+            val (cvvValidationResult) = cvvResult
+            cvvValidationResult.complete
+        } ?: true
+    }
+
+    override fun onUpdatePAN(pan: String) {
+    }
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/main/java/com/worldpay/access/checkout/AccessCheckoutCVV.kt]: I agree, I will do this.

File: checkout/access-checkout-android/src/main/java/com/worldpay/access/checkout/AccessCheckoutCVV.kt [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        } else {
+            validateCompleteCVV(cardValidator, cvvView.getInsertedText())
+        }
+    }
+
+    override fun isValid(): Boolean {
+        val cvv = cvvView.getInsertedText()
+
+        return cardValidator?.let {
+            val cvvResult = it.validateCVV(cvv, null)
```
[the remark this answers is not in the document]
Comment (juned-miah) [reply checkout/access-checkout-android/src/main/java/com/worldpay/access/checkout/AccessCheckoutCVV.kt]: done

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt (hunk 2 above)
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: Come to think of it, no we don't. This test was added to ensure we do not use the cardListener as a null, avoiding the NPE. However in Kotlin, it fails to compile anyway if you try to do this. It is more  of a test case if this was in Java so I can remove this test.

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt (hunk 4 above)
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: Done

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt (hunk 3 above)
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: Yes this is correct, the client can easily provide a factory that returns a null `cvvLengthFilter` and so this test is the ensure that we can handle this.

File: checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt (hunk 1 above)
Comment (juned-miah) [reply checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt]: good idea, done this now

**Pull request merged**